### DISCUSSION TITLE
validation 2.0 - validation of shiny apps

### DISCUSSION THREAD

**Discussion by harveyl888** [rinpharma/positconf-roundtables-2023#9] [Ideas]
There's been a lot of work and ideas generated on R package validation.  A follow-up topic is on validation of shiny apps.  Colleagues at Novartis have been working on new workflows for shiny app validation as the existing route required validation through older complex IT processes (along with the huge paper trail that they generate).  How are companies attacking this internally?  Are there synergies and can we learn from each other?

**Comment by epijim** [comment]
plus one this - I would maybe step this back slightly to be about rigor and traceability in shiny, as whether validation is applicable to shiny is something that I don't think all companies agree on 

As an example we validate the static packages which contain the shiny modules with the scientific logic, but we don't validate the UI where someone has chained together those modules into an app (that final app is QC'd like study code). But I guess I'm highlighting I agree this is an awesome topic to have a round table on.

**Reply by harveyl888** [reply]
Yep.  It would be interesting to hear how each company is addressing this and what level of detail is required.

**Comment by MikeKSmith-Pfizer** [comment]
We (Pfizer) have recently proposed a framework for production Shiny apps, where the app is based on existing functions which have documentation and testthat cases. This allows the functions to be used outside the app, and to be "validated" through test cases. We then build the app around those functions, and use {shinytest} to test the UI and server logic for the app. By capturing user inputs / reactives we can then run a template script using the same functions with the captured reactives as inputs to recreate the Shiny output via a simple to validate script. 

That way, the Shiny app is a UI for the functions / package used and the definitive production run is from the template script, outside of Shiny. 

I agree that this is a good topic for discussion since I guess every company deploying Shiny has the same question...

**Reply by dgkf-roche** [reply]
This tracks pretty closely with what we do at Roche - Ask that devs bundle as much of the app business logic as possible in a package where it can be more directly tested, and then use shinytest for the UI.

The bigger hurdle that we've encountered is the difficulty of locking down an environment served via rsconnect. rsconnect has some pretty opaque mechanisms for fetching dependencies for an app, which has made reproducibility difficult. Pulling explicit versions of dependencies (especially internal dependencies) is quite difficult, and because of this our testing environment can only approximate the rsconnect server environment. We couch all our shiny app validation with caveats for these behaviors, and for now are quite conservative with the risk tolerance for shiny app use cases.

**Reply by MikeKSmith-Pfizer** [reply]
Your point about environment is well taken. We deploy containers in Connect that are the same containers deployed in Workbench - if users publish from Workbench to Connect then (in theory) the container won't need to install any other packages, it can reuse what's already available. 

What I'd like Posit to consider is full use of {renv} with Connect so that when we publish / deploy with a {renv} lockfile then we can predict with more certainty what packages and versions are being used on Connect.
